networking/src/Ntp/Packet.hs

```diff
@@ -4,92 +4,142 @@
-ntpPacketSize = 48
-
-exp2'32 :: Integer
-exp2'32 = 4294967296
-
-ntpToRealMcs :: Word32 -> Word32 -> Microsecond
-ntpToRealMcs integerSec fracSec = fromMicroseconds $
-       (fromIntegral integerSec - ntpTimestampDelta) * 1000000
-      + ((fromIntegral fracSec * 1000000) `div` exp2'32)
-
-realMcsToNtp :: Microsecond -> (Word32, Word32)
-realMcsToNtp (toMicroseconds -> mcs) =
-    let integerSec = (mcs `div` 1000000) + ntpTimestampDelta
-        fracSec    = ((mcs * exp2'32) `div` 1000000)
-    in  (fromIntegral integerSec, fromIntegral fracSec)
+    { ntpParams       :: Word8        -- ^ some magic parameters
+    , ntpPoll         :: Int8         -- ^ poll delay between requests
+    , ntpOriginTime   :: Microsecond  -- ^ when server sent reply
+    , ntpReceivedTime :: Microsecond  -- ^ when server got request
+    , ntpTransmitTime :: Microsecond  -- ^ when client sent request
+    } deriving (Show, Eq)
 
 instance Binary NtpPacket where
     put NtpPacket{..} = do
         putWord8 ntpParams
+        putWord8 0
+        putInt8 ntpPoll
+        putWord8 0
 
-        -- since it's sent only by client, initialize only `transmitTime`
-        replicateM_ 3 $ putWord8 0
-        replicateM_ 9 $ putWord32be 0
+        replicateM_ 5 $ putWord32be 0
+
+        mapM_ putWord32be $ realMcsToNtp ntpOriginTime ^.. each
+        mapM_ putWord32be $ realMcsToNtp ntpReceivedTime ^.. each
         mapM_ putWord32be $ realMcsToNtp ntpTransmitTime ^.. each
 
     get = do
         ntpParams <- getWord8
         _         <- getWord8
-        ntpPoll   <- fromIntegral <$> getInt8
+        ntpPoll   <- getInt8
         _         <- getWord8
-        replicateM_ 5 getWord32be
+
+        -- skip 5 @'Word32'@ words
+        skip 20
+
         ntpOriginTime   <- getTimestamp
         ntpReceivedTime <- getTimestamp
         ntpTransmitTime <- getTimestamp
         return NtpPacket{..}
       where
         getTimestamp = ntpToRealMcs <$> getWord32be <*> getWord32be
 
-mkCliNtpPacket :: MonadIO m => m NtpPacket
-mkCliNtpPacket = do
+-- |
+-- NTP timestamp start in 1.1.1900, i.e. 70 years before UNIX epoch.
+-- references:
+--  * https://tools.ietf.org/html/rfc5905#section-6
+--  * https://tools.ietf.org/html/rfc5905#appendix-A.4
+ntpTimestampDelta :: Integer
+ntpTimestampDelta = 2208988800
+
+-- |
+-- We only need first 48 bytes of a packet:
+-- reference: https://tools.ietf.org/html/rfc5905#section-7.3
+ntpPacketSize :: Int
+ntpPacketSize = 48
+
+-- |
+-- For pairs @(x, y) :: (Word32, Word32)@ with @y \`mod\` 4294 == 0@ it is
+-- be right inverse of @'realMsgToNtp'@.   In general it is not injective (for that
+-- we'd need to use @'Picosecond'@ instead of @'Microsecond'@).
+ntpToRealMcs :: Word32 -> Word32 -> Microsecond
+ntpToRealMcs sec frac =
+    let -- microseconds
+        secMicro :: Integer
+        secMicro = (fromIntegral sec - ntpTimestampDelta) * 1000000
+        -- Each fraction resolves to 232 pico seconds (`232 ≈ 10000000/4294`)
```

Comment:
There's an extra `0` in the numerator.

The explanation could be better. We divide `1` second into `2 ^ 32` parts, giving `2.3283064365386963e-10` as the quantum. A picosecond is `10e-12` of a second, so this is `232` picoseconds.

Reply:
Sorry, I put this comment under wrong thread.